### PR TITLE
Make dependency on Boost::program_options optional

### PR DESCRIPTION
**Pull request by uulm-janbaudisch**
Currently, the C library interface depends on `Boost::program_options`, which might not be needed when using the library.

With this, the default behaviour is not changed, but the dependency can be removed by setting `MT_KAHYPAR_IO_CMD` to `false` and therefore not including `io/command_line_options` (the only part in the library where `Boost::program_options` is used) in the library.

### CODE REVIEW

File: lib/CMakeLists.txt [hunk 1]
```diff
@@ -1,7 +1,7 @@
 include(GNUInstallDirs)
 
 add_library(mtkahypar SHARED libmtkahypar.cpp)
-target_link_libraries(mtkahypar ${Boost_LIBRARIES} TBB::tbb TBB::tbbmalloc_proxy)
+target_link_libraries(mtkahypar TBB::tbb TBB::tbbmalloc_proxy)
```
[the remark this answers is not in the document]
Comment (uulm-janbaudisch) [reply lib/CMakeLists.txt:5]: Yes, will be linked now.

**Comment by larsgottesbueren** [comment]
The Windows build is failing in the CI. Can you please take another look.

**Review by larsgottesbueren — APPROVED** [review]

**Review by kittobi1992 — CHANGES_REQUESTED** [review]
Many thanks for bringing this up.

There are a couple more places where `mt-kahypar/io/command_line_options.h` is used, e.g., in the python interface (see `module.cpp`) and in some unit tests (see `context_test.cc.in` and `initial_partitioner_test.cc`). 

I would prefer having not too much `#ifdef` in the code. We can implement most of the change in the cmake files. For example, if boost is not available it makes no sense to have a target for our binary. Similarly, for the unit tests. Just exclude the tests that use `command_line_options.h` in the corresponding cmake files. 

The only place where we need `#ifdef` is in the C and Python interface to remove the functions that read a configuration file.

**Comment by uulm-janbaudisch** [comment]
I made the same change to the Python interface as I did with the library and excluded the other files.
It builds and the resulting libraries do not depend on `Boost::program_options`. However, I did not do further testing.

File: lib/libmtkahypar.cpp [hunk 2]
```diff
@@ -41,11 +41,14 @@
 #include "mt-kahypar/parallel/stl/scalable_vector.h"
 #include "mt-kahypar/io/hypergraph_factory.h"
 #include "mt-kahypar/io/hypergraph_io.h"
-#include "mt-kahypar/io/command_line_options.h"
 #include "mt-kahypar/macros.h"
 #include "mt-kahypar/utils/cast.h"
 #include "mt-kahypar/utils/delete.h"
 
+#if !MT_KAHYPAR_DISABLE_BOOST
```
Comment (kittobi1992) [inline lib/libmtkahypar.cpp:48]:
```suggestion
#ifndef MT_KAHYPAR_DISABLE_BOOST
```

File: lib/libmtkahypar.cpp [hunk 3]
```diff
@@ -85,6 +88,8 @@ void mt_kahypar_free_context(mt_kahypar_context_t* context) {
   delete reinterpret_cast<Context*>(context);
 }
 
+#if !MT_KAHYPAR_DISABLE_BOOST
```
Comment (kittobi1992) [inline lib/libmtkahypar.cpp:91]:
```suggestion
#ifndef MT_KAHYPAR_DISABLE_BOOST
```

File: python/module.cpp [hunk 4]
```diff
@@ -43,12 +43,15 @@
 #include "mt-kahypar/partition/metrics.h"
 #include "mt-kahypar/partition/partitioner.h"
 #include "mt-kahypar/partition/mapping/target_graph.h"
-#include "mt-kahypar/io/command_line_options.h"
 #include "mt-kahypar/io/hypergraph_factory.h"
 #include "mt-kahypar/io/hypergraph_io.h"
 #include "mt-kahypar/utils/cast.h"
 #include "mt-kahypar/utils/randomize.h"
 
+#if !MT_KAHYPAR_DISABLE_BOOST
```
Comment (kittobi1992) [inline python/module.cpp:51]:
```suggestion
#ifndef MT_KAHYPAR_DISABLE_BOOST
```

File: python/module.cpp [hunk 5]
```diff
@@ -261,10 +264,13 @@ PYBIND11_MODULE(mtkahypar, m) {
         }
       }, "Loads a preset for partitioning (DETERMINISTIC, LARGE_K, DEFAULT or QUALITY)",
       py::arg("preset type"))
+#if !MT_KAHYPAR_DISABLE_BOOST
```
Comment (kittobi1992) [inline python/module.cpp:267]:
```suggestion
#ifndef MT_KAHYPAR_DISABLE_BOOST
```

File: python/module.cpp [hunk 6]
```diff
@@ -261,10 +264,13 @@ PYBIND11_MODULE(mtkahypar, m) {
...
+#include "mt-kahypar/io/command_line_options.h"
```
Comment (kittobi1992) [inline python/module.cpp:268]: This should be removed

File: .github/workflows/master_ci.yml [hunk 7]
```diff
@@ -216,7 +216,7 @@ jobs:
           mkdir build
           cd build
           export CMAKE_GENERATOR="MSYS Makefiles"
-          /mingw64/bin/cmake .. -DCMAKE_BUILD_TYPE=RELEASE -DKAHYPAR_DOWNLOAD_BOOST=On -DKAHYPAR_CI_BUILD=ON -DKAHYPAR_ENABLE_HIGHEST_QUALITY_FEATURES=OFF -DKAHYPAR_ENABLE_LARGE_K_PARTITIONING_FEATURES=OFF
```
Comment (kittobi1992) [inline .github/workflows/master_ci.yml:219]: Why is `-DKAHYPAR_DOWNLOAD_BOOST=On` removed here?

Comment (uulm-janbaudisch) [reply .github/workflows/master_ci.yml:219]: Because the Boost MSYS2 package is already downloaded prior and therefore there is no need to build it again.

**Review by kittobi1992 — APPROVED** [review]
LGTM! Please squash&merge if you merge this PR

**Comment by juliannz** [comment]
While working on #177, I noticed that the `MT_KAHYPAR_DISABLE_BOOST` option does not keep you from being forced to have boost installed (downloaded).

With extending [this](https://github.com/kahypar/mt-kahypar/blob/d784c914a5c5a85abc9747eaacefa5d6f9905cce/CMakeLists.txt#L174) to `elseif(NOT MT_KAHYPAR_DISABLE_BOOST)`, you could build the library without any boost installation. Would this be desirable behaviour?

**Comment by kittobi1992** [comment]
Yes, this should be the desired behaviour. I will merge your PR if CI is completed.

**Comment by N-Maas** [comment]
@uulm-janbaudisch Unfortunately, the `MT_KAHYPAR_DISABLE_BOOST` is no longer available because the changes in #196 require boost even for loading the presets. However, if the motivation for this is avoiding link dependencies, you can just link boost statically now.

**Pull request merged**